move-mutator/src/coverage.rs

```diff
@@ -0,0 +1,221 @@
+                    all_uncovered_spans.insert(associated_fn_name, uncovered_spans);
+                }
+            });
+
+        trace!("all uncovered spans: {all_uncovered_spans:?}");
+        self.all_uncovered_spans = all_uncovered_spans;
+        Ok(())
+    }
+
+    /// Check if the location is covered by the unit test.
+    pub(crate) fn check_location(&self, associated_fn_name: String, loc: &Loc) -> bool {
+        let span = loc.span();
+
+        let Some(UncoveredSpans(spans)) = self.all_uncovered_spans.get(&associated_fn_name) else {
+            trace!("location has coverage since {associated_fn_name} has full coverage");
+            return true;
+        };
+
+        for uncovered_span in spans {
+            // Skip all early spans.
+            if span.start() >= uncovered_span.end() {
+                continue;
+            }
+
+            // Check if the span starts earlier than the uncovered span.
+            if uncovered_span.start() > span.start() {
+                break;
+            }
+
+            // If the span goes beyond the uncovered span, we are done here.
+            if uncovered_span.end() < span.end() {
+                break;
+            }
+
+            trace!("{associated_fn_name} has no coverage for the given location");
+            return false;
+        }
+
+        trace!("{associated_fn_name} has coverage for the given location");
+        true
+    }
+}
+
+#[derive(Debug)]
+struct UncoveredSpans(Vec<Span>);
+
+impl UncoveredSpans {
+    /// Create a new [`UncoveredSpans`].
+    fn new(func_src_cov: FunctionSourceCoverage, file_hash: FileHash, file_contents: &str) -> Self {
+        let merged_spans = merge_spans(file_hash, func_src_cov);
+        let optimized_spans = merge_spans_after_removing_whitespaces(merged_spans, file_contents);
+        Self(optimized_spans)
+    }
+}
+
+// This is based on the same function in aptos-core, but that one needs tweaking - it doesn't merge
+// spans properly in all cases, but that's not noticeable in the coverage output. We could port the
+// logic from here in their repo and then expose that function and use that one directly from
+// there.
+/// Efficiently merge spans.
+fn merge_spans(file_hash: FileHash, cov: FunctionSourceCoverage) -> Vec<Span> {
+    if cov.uncovered_locations.is_empty() {
+        return vec![];
+    }
+
+    let mut spans: Vec<_> = cov
+        .uncovered_locations
+        .iter()
+        .filter(|loc| loc.file_hash() == file_hash)
+        .map(|loc| Span::new(loc.start(), loc.end()))
+        .collect();
+    if spans.is_empty() {
+        return vec![];
+    }
+    spans.sort();
+
+    let mut all_spans = Vec::new();
+    let mut curr = spans.remove(0);
+
+    for span in spans {
+        if curr.end() >= span.start() {
+            curr = curr.merge(span);
+        } else {
+            all_spans.push(curr);
+            curr = span;
+        }
+    }
+
+    all_spans.push(curr);
+    all_spans
```

Comment:
Could we do something like that?

```suggestion
cov.uncovered_locations
        .iter()
        .filter(|loc| loc.file_hash() == file_hash)
        .map(|loc| Span::new(loc.start(), loc.end()))
        .collect::<Vec<_>>()
        .windows(2)
        .fold(vec![], |mut acc, spans| {
            let [curr, next] = spans else { return acc };
            if curr.end() >= next.start() {
                acc.push(curr.merge(*next));
            } else {
                acc.push(*curr);
            }
            acc
        })
```

Reply:
Very cool! Thanks!
Done